### PR TITLE
Make typst-cli crate usable as library

### PR DESCRIPTION
**Pull request by zinifier**
Like others in #4452 i would like to use typst as library. This is the simplest way i found to do it, turning typst_cli into a library.

### CODE REVIEW

File: crates/typst-cli/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+
+thread_local! {
+    /// The CLI's exit code.
+    pub static EXIT: Cell<ExitCode> = const { Cell::new(ExitCode::SUCCESS) };
+}
+
+/// The parsed commandline arguments.
+pub static ARGS: Lazy<CliArguments> = Lazy::new(CliArguments::parse);
```
Comment (tingerrr) [inline crates/typst-cli/src/lib.rs:38]: This is implicitly accessed in various parts of the library and will exit if the cli arguments don't match.

File: crates/typst-cli/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+pub mod fonts;
+pub mod init;
+pub mod package;
+pub mod query;
+pub mod terminal;
+pub mod timings;
+#[cfg(feature = "self-update")]
+pub mod update;
+pub mod watch;
+pub mod world;
```
Comment (tingerrr) [inline crates/typst-cli/src/lib.rs:13]: Exposing all of these comes with a big semver hazard, if any of these APIs which are specific to typst-cli alone like `update` are changed, a new major version of the library would be required.

In theory, no one in their right mind would use the `update` module, but this is not enough of an argument to ignore the semver hazard I'm afraid.

**Review by tingerrr — COMMENTED** [review]
I'm afraid turning typst-cli into a library crate just like this will not be enough. The implicit dependency on typst-cli specific arguments (`ARGS`) must be removed from all publically exported functions.

File: crates/typst-cli/src/lib.rs (hunk 1 above)
Comment (zinifier) [reply crates/typst-cli/src/lib.rs:38]: I think it's only used in terminal.rs for color management, and download.rs for cert management, and main.rs of course. It works fine when calling compile_once from outside typst.

**Comment by zinifier** [comment]
> I'm afraid turning typst-cli into a library crate just like this will not be enough.

Yes sorry i should have made it more clear: i'm not saying this is the good way to do it. I'm just showing that's what i did that works for me. Feel free to close the pull request if you have better plans.

File: crates/typst-cli/src/lib.rs (hunk 1 above)
Comment (tingerrr) [reply crates/typst-cli/src/lib.rs:38]: But that's the point, a consumer might try to use download and then get a failure about their cli arguments being incorrect when they might not be making a cli program at all.

**Comment by tingerrr** [comment]
> Yes sorry i should have made it more clear: i'm not saying this is the good way to do it. I'm just showing that's what i did that works for me. Feel free to close the pull request if you have better plans.

I see, I can't manage pull requests, but I think it would be good to turn this into a draft (see the upper right next to reviewers) to clearly indicate that this is more of an example than a finished proposed change to the repo.

**Comment by laurmaedje** [comment]
I'm on board with a `typst-kit` crate with a proper API to help with downstream tooling (as discussed in https://github.com/typst/typst/issues/4452, @tingerrr is already working on it). I think that's the cleaner solution, so I'll close this. Thanks still!

**Pull request closed without merging**